Fwupd download scripts

Asiderr opened the pull request:
Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
src/: Add download scripts

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
test/: Add test file

src/: Rename fwupd-receive-updates and create qubes_fwupdmgr.py

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Add metadata methods,

test_qubes_fwupdmgr.py: Add tests for metadata methods
README.md: Add testing instruction

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Add firmware update UI

qubes_fwupdmgr.py: Add get-device and get-updates functionalities
qubes_fwupdmgr.py: Provide firmware update method, add main function
test_qubes_fwupdmgr.py: Add tests for fw update UI
fwupd_receive_updates.py: Copy downloaded update to dom0 update dir

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Add _get_device info method

qubes_fwupdmgr.py: Add help method, add clean_cache method
qubes_fwupdmgr.py: Fix style
fwupd_receive_updates.py: Fix style
test_qubes_fwupdmgr.py: Add tests for output methods
logs/: Add logs for output tests

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
setup: Create setup script

README.md: fix typo
fwupd-dom0-update: Change updateVM script directory

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Fix typos

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
test_qubes_fwupdmgr.py: Fix typo in metadata refresh assertion

qubes_fwupdmgr.py: Fix typo in firmware update method

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
fwupd-dom0-update: Fix help message

fwupd_receive_updates.py: Fix path concatenation
qubes_fwupdmgr.py: Fix path concatenation

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Fix typo in path concatenation

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
src:/ Add docstrings to the methods

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Add dmi verification

logs/: Add firmware.metainfo.xml test files
fwupd_receive_updates.py: Fix typos in docstrings

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Fix non parameter call

qubes_fwupdmgr.py: Replace exception with exit(1)

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Install top version instead of all

setup: Remove symlink before creating new one
test_qubes_fwupdmgr.py: Fix test_parse_parameters

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Add downgrade functionality

test_qubes_fwupdmgr.py: Test downgrade functionality
fwupd_logs.py: Move fwupd logs to separate file

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Add help information and downgrade argument

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Fix typo

help.log: Add update log

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

macpijan commented on README.md:
````diff
@@ -1,2 +1,10 @@
+## Testing
+
+```
+cd /usr/share/fwupd-qubes/
````
It assumes that package is already installed there? There is no installation documentation as far I can see.
For testing info in README  I would expect something like:
- clone
- execute this command to run tests
- maybe example command to run tests

But maybe it is too tightly coupled with the QubesOS itself?

For unit tests, it would be nice (at lest for some portion of them) to be usable not only in the Qubes. Ideally, a docker container for automatic testing could be setup (if there are no major blockers, like usage of fwupd in container). But that can be left as an extra work after we finish the major requirements.

Asiderr replied:
I've included the installation process and usage in README.  

> For unit tests, it would be nice (at lest for some portion of them) to be usable not only in the Qubes.

It is already implemented. Tests that require Qubes are omitted if you are testing on another OS. (I've also added a few words in README about this topic).

macpijan replied:
Coll. Verified that.

macpijan commented on setup:
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
```
Ok, I can see a setup script here. When it is called? A README might explain that.

Asiderr replied:
The setup script is used for development proposes. I've added explanation how to use it in the README.

macpijan commented on setup:
```diff
@@ -0,0 +1,30 @@
+
+# Create symlink
+rm /bin/qubes-fwupdmgr
+ln -s /usr/share/fwupd-qubes/src/qubes_fwupdmgr.py /bin/qubes-fwupdmgr
```
Is it how the others qubes python script installed?

Asiderr replied:
In a final solution we will use [qubes-builder](https://github.com/QubesOS/qubes-builder/) to install the package (it will be added in further development). This script is for only development proposes.

macpijan commented on src/fwupd-download-updates.sh:
```diff
@@ -0,0 +1,87 @@
+#!/bin/bash
+
+FWUPD_UPDATES_DIR=/home/user/.cache/fwupd
```
I like to have a common .sh file with common set of variables and source it at the beginning of the script to have the common variables defined only in one place. But maybe this is fine here since it is used only twice there.

Another thing: you are using `/home/user/.cache/fwupd` path three times in different files and the variable name is different each time with a slightly different meaning. Let's focus on what this path (and possibly others) really is and be consistent.

Asiderr replied:
> I like to have a common .sh file with common set of variables and source it at the beginning of the script to have the common variables defined only in one place.

fwupd-download-updates.sh and fwupd-dom0-update are in different VMs, so there would be no profit in this case.

> Let's focus on what this path (and possibly others) really is and be consistent.

I fixed that problem.

macpijan commented on src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,281 @@
+            )
+
+    def _check_domain(self, updatevm):
+        # Check if updateVM is allowed to provide update files
```
This comments seems redundant now when the docstring is here.

Asiderr replied:
Deleted.

macpijan commented on src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,281 @@
+        file_path -- absolute path to file
+        sha -- SHA1 checksum of the file
+        """
+        # Count sha1 checksum
```
calcualate or compute fits probably better (as in your docstring above). Maybe it is just redundant.

Asiderr replied:
Deleted.

macpijan commented on src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,281 @@
+            exit(1)
+
+    def _verify_received(self, files_path, regex_pattern):
+        """Checks if sent files are expected.
```
Can we rephrase that/clarify that somehow?
IIUC, we just check here if the filename path matches the regex we expect.

Asiderr replied:
> IIUC, we just check here if the filename path matches the regex we expect.

That's right. I've clarified that.

macpijan commented on src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,281 @@
+        if args is None:
+            raise ValueError("Creating directories failed, no paths given.")
+        for file_path in args:
+            if not path.exists(file_path):
```
What if it exists, but maybe the permissions are not set correctly? How critical is that? Maybe we shall check that or print some information to the user that the directory should be moved.

Asiderr replied:
We have sudo rights here, it is not so critical. It may cause data lose in case of cleaning cache. I've set warning here.

macpijan commented on src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,281 @@
+                os.chmod(file_path, 0o0775)
+
+    def _extract_archive(self, archive_path, output_path):
+        """Extracts archive file to specified directory.
```
```suggestion
        """Extracts archive file to the specified directory.
```

Asiderr replied:
Fixed.

macpijan commented on src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,281 @@
+        """Compares computed SHA1 checksum with `sha` parameter.
+
+        Keyword arguments:
+        file_path -- absolute path to file
```
```suggestion
        file_path -- absolute path to the file
```

Asiderr replied:
Fixed.

macpijan commented on src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,281 @@
+
+        Keyword arguments:
+        archive_path -- absolute path to archive file
+        output_path -- absolute output directory
```
```suggestion
        output_path -- absolute path to the output directory
```

Asiderr replied:
Fixed.

macpijan commented on src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,281 @@
+        archive_path -- absolute path to archive file
+        output_path -- absolute output directory
+        """
+        cmd_extract = [
```
Is there a Python module for extracting it, rather than using shell command?

Asiderr replied:
There is python module - [cabarchive](https://github.com/hughsie/python-cabarchive), but it uses the same [subprocess](https://github.com/hughsie/python-cabarchive/blob/master/test.py#L32) to extract archive.

macpijan commented on src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,281 @@
+        Keyword argument:
+        file_path -- absolute path to inspected file
+        """
+        cmd_gpg = [
```
There must be Python binding for gpg. Generally, it should be preferred to use Pyhton libraries rather than shell commands and parsing outputs. But let's check how it is done (maybe you have already) in others Qubes python scripts and be consistent with that.

Asiderr replied:
Packages installed in dom0 are verified with [rpm subprocess](https://github.com/3mdeb/qubes-core-admin-linux/blob/a3098d517f9f85af4ec02db4077ecf2192f340a5/dom0-updates/qubes-receive-updates#L95).

> Generally, it should be preferred to use Pyhton libraries rather than shell commands and parsing outputs.

When we are using the shell commands, we don't have to provide any additional python modules to dom0. This is a security aspect that should be taken into account.

> There must be Python binding for gpg.

There is python module https://pythonhosted.org/python-gnupg/, but it is wrapper for gpg command which uses [gpg subprocess](https://github.com/vsajip/python-gnupg/blob/master/gnupg.py#L934)

macpijan replied:
Let's stick with the Qubes way of doing things and keep it it this way then. The security aspect might be a thing.

macpijan commented on src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,281 @@
+            )
+
+    def _gpg_verification(self, file_path):
+        """Inspects GPG signature.
```
```suggestion
        """Verifies GPG signature.
```

Asiderr replied:
Fixed.

macpijan commented on src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,281 @@
+            )
+
+    def handle_fw_update(self, updatevm, shasum, filename):
+        """Handles copying firmware update archives from the updateVM.
```
Maybe just `Copies files` ? Try to be consistent in style in the comments / docs.

Asiderr replied:
Fixed.

macpijan commented on src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,281 @@
+                'Domain updateVM sent not signed firmware: ' + file_path
+            )
+
+    def handle_fw_update(self, updatevm, shasum, filename):
```
In some functions it is `shasum` in others `sha`. Let's be consistent if this represents the same thing.

Asiderr replied:
Fixed.

macpijan commented on src/qubes_fwupdmgr.py:
```diff
@@ -0,0 +1,502 @@
+        if p.returncode != 0:
+            raise Exception("fwudp-qubes: Metadata update failed")
+        if not os.path.exists(FWUPD_DOM0_METADATA_FILE):
+            raise ValueError("Metadata signature does not exist")
```
is it really ValueError? It applies to other lines as well. I think, most of the time is not really one (as defined here: https://docs.python.org/3/library/exceptions.html#ValueError). I suggest to pick another one or just throw regular exception.

Asiderr replied:
I've verified all exceptions and I've fixed wrong ones.

macpijan commented on src/qubes_fwupdmgr.py:
```diff
@@ -0,0 +1,502 @@
+        """Initializes downloading firmware upadate archive.
+
+        Keywords arguments:
+        url -- url path to firmware upadate archive
```
```suggestion
        url -- url path to the firmware upadate archive
```

Asiderr replied:
Fixed.

macpijan commented on src/qubes_fwupdmgr.py:
```diff
@@ -0,0 +1,502 @@
+        url -- url path to firmware upadate archive
+        sha -- SHA1 checksum of the firmware update archive
+        """
+        name = url.replace("https://fwupd.org/downloads/", "")
```
Maybe this should be a constant as well? I can see it being used a few times.

Asiderr replied:
You're right. Done.

macpijan commented on src/qubes_fwupdmgr.py:
```diff
@@ -0,0 +1,502 @@
+                    "Otherwise choose a device number: "
+                )
+                if choice == 'N' or choice == 'n':
+                    return 99
```
Can we define a dict with possible exit codes and name them somehow, rather than hardocding integers

Asiderr replied:
Done.

macpijan commented on src/qubes_fwupdmgr.py:
```diff
@@ -0,0 +1,502 @@
+        path - absolute path to firmware update archive
+        """
+        cmd_install = [
+            "/bin/fwupdmgr",
```
maybe constant as well

Asiderr replied:
Done.

macpijan commented on src/qubes_fwupdmgr.py:
```diff
@@ -0,0 +1,502 @@
+    def _read_dmi(self):
+        """Reads BIOS information from DMI."""
+        cmd_dmidecode = [
+            "dmidecode",
```
Have you checked for python module?

Asiderr replied:
There is external wrapper for dmidecode. It uses similar [subprocess](https://github.com/zaibon/py-dmidecode/blob/master/dmidecode.py#L210)

Commit:
README.md: Add installation and usage description

fwupd_receive_updates.py: Fix typos and exception problems
fwupd-dom0-update: Unify variables
fwupd-download-updates.sh: Unify variables
qubes_fwupdmgr.py: Fix typos, and exception problems
qubes_fwupdmgr.py: Add exit code dictionary
test_qubes_fwupdmgr.py: Fix paths

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
test_qubes_fwupdmgr.py: Fix typo

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Fix N choice for downgrade

test_qubes_fwupdmgr.py: Fix downgrade test and add update frimware test
fwupd_receive_updates.py: Delete blank line

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Commit:
qubes_fwupdmgr.py: Fix double user input in downgrade

test_qubes_fwupdmgr.py: Fix downgrade test and update firmware test

Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>
